packages/preferences/src/browser/abstract-resource-preference-provider.ts

```diff
@@ -83,61 +105,86 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
-                newContent = jsoncparser.applyEdits(content, edits);
+                const { insertSpaces, tabSize, defaultEOL } = textModel.getOptions();
+                for (const edit of jsoncparser.modify(content, path, value, {
+                    formattingOptions: {
+                        insertSpaces,
+                        tabSize,
+                        eol: defaultEOL === monaco.editor.DefaultEndOfLine.LF ? '\n' : '\r\n'
+                    }
+                })) {
+                    const start = textModel.getPositionAt(edit.offset);
+                    const end = textModel.getPositionAt(edit.offset + edit.length);
+                    editOperations.push({
+                        range: monaco.Range.fromPositions(start, end),
+                        text: edit.content || null,
+                        forceMoveMarkers: false
+                    });
+                }
+            } else {
+                editOperations.push({
+                    range: textModel.getFullModelRange(),
+                    text: null,
+                    forceMoveMarkers: false
+                });
             }
-            await resource.saveContents(newContent);
+            await this.workspace.applyBackgroundEdit(this.model, editOperations);
+            return true;
         } catch (e) {
-            const message = `Failed to update the value of ${key}.`;
-            this.messageService.error(`${message} Please check if ${resource.uri.toString()} is corrupted.`);
-            console.error(`${message} ${e.toString()}`);
+            const message = `Failed to update the value of '${key}' in '${this.getUri()}'.`;
+            this.messageService.error(`${message} Please check if it is corrupted.`);
+            console.error(`${message}`, e);
             return false;
         }
-        await this.readPreferences();
-        return true;
     }
 
     protected getPath(preferenceName: string): string[] | undefined {
         return [preferenceName];
     }
 
-    protected loaded = false;
-    protected async readPreferences(): Promise<void> {
-        const newContent = await this.readContents();
-        this.loaded = newContent !== undefined;
-        const newPrefs = newContent ? this.getParsedContent(newContent) : {};
-        this.handlePreferenceChanges(newPrefs);
-    }
-
-    protected async readContents(): Promise<string | undefined> {
+    /**
+     * It HAS to be sync to ensure that `setPreference` returns only when values are updated
+     * or any other operation modifying the monaco model content.
+     */
+    protected readPreferences(): void {
+        const model = this.model;
+        if (!model || model.dirty) {
+            return;
+        }
         try {
-            const resource = await this.resource;
-            return await resource.readContents();
-        } catch {
-            return undefined;
+            let preferences;
+            if (model.valid) {
```

Comment:
@AlexTugarev since we don't reset model to empty value when resource is gone, i have to add such guard here that preferences don't get created from invalid model